src/bun.js/bindings/URL.zig

```diff
@@ -121,6 +135,63 @@ pub const URL = opaque {
         if (len == 0) return null;
         return slice[0..len];
     }
+
+    /// Utilities for interacting with URL components as slices.
+    pub fn sliceTools(url: *URL) SliceTools {
+        return .{ .url = url };
+    }
+
+    pub const SliceTools = struct {
+        url: *URL,
+
+        pub fn protocol(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.protocol();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+
+        pub fn host(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.host();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+
+        pub fn pathname(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.pathname();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+
+        pub fn hash(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.hash();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+
+        pub fn href(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.href();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+
+        pub fn fragmentIdentifier(self: SliceTools, allocator: std.mem.Allocator) ![]const u8 {
+            const str = self.url.fragmentIdentifier();
+            defer str.deref();
+            const utf8 = str.toUTF8(allocator);
+            defer utf8.deinit();
+            return allocator.dupe(u8, utf8.slice());
+        }
+    };
```

Comment:
these can be deleted and replaced with `str.toOwnedSlice(allocator)` at each callsite